### PR TITLE
Initial generator project

### PR DESCRIPTION
**Pull request by lilleyse**
Generates tilesets like those used in Cesium's spec/data folder and in 3d-tiles-samples.

To-do:

* [x] Add a few more missing tilesets
* [ ] Open a PR into Cesium with the newly converted tilesets - check that everything still works as expected.
* [x] Update project files a bit
* [ ] BatchTableHierarchy generator - I have some initial code for this but I would like to clean it up, doesn't need to be this PR
* [x] Use correct date in CHANGES.md
* [x] Write some basic tests. What's the best way to test a bin script?
* [ ] Open PR in 3d-tiles-samples with newly converted tilesets and README note.
* [x] Run any glb's in the data folder through gltf-pipeline if needed. Still preserve _3DTILESDIFFUSE semantic.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Is it worth adding some super simple tests, even if they just have a set of known tilesets, runs the most typical commands, and then confirms that the output matches?

**Comment by pjcozzi** [comment]
Please add third-party npm dependencies that we ship for this package and the other packages in this repo to the LICENSE.md file.  This should be maintained as we add/remove packages.

**Comment by pjcozzi** [comment]
Can you add some doc to the README.md?

**Comment by pjcozzi** [comment]
> Open a PR into Cesium with the newly converted tilesets - check that everything still works as expected.

Please include a script and instructions for this.

**Comment by pjcozzi** [comment]
> Can you add some doc to the README.md?

We don't have to overbuild this project so, the instruction might be "changes these variables in this file that is commented" if you don't already have a CLI.

**Comment by pjcozzi** [comment]
Are any of these script used to generate tilesets in the 3d-tiles-samples repo?  If so, please open a PR in that repo to update the README.md with a link here and instructions on how to reproduce the tileset.

File: generator/lib/createPointCloudTile.js [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+ * @param {Boolean} [options.generateNormals=false] Generate per-point normals.
+ * @param {Boolean} [options.octEncodeNormals=false] Apply oct16p encoding on the point normals.
+ * @param {Boolean} [options.quantizePositions=false] Quantize point positions so each x, y, z takes up 16 bits rather than 32 bits.
+ * @param {Boolean} [options.batched=false] Group points together with batch ids and generate per-batch metadata. Good for differentiating different sections of a point cloud. Not compatible with perPointProperties.
+ * @param {Boolean} [options.perPointProperties=false] Generate per-point metadata.
+ * @param {Boolean} [options.relativeToCenter=true] Define point positions relative-to-center.
+ *
+ * @returns {Object} An object containing the pnts buffer and batch table JSON.
+ */
+
```
Comment (pjcozzi) [inline generator/lib/createPointCloudTile.js]: Throughout, I've noticed empty whitespace like this if you care.

File: generator/lib/createPointCloudTile.js [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+    var theta = CesiumMath.nextRandomNumber() * 2 * Math.PI;
+    var phi = CesiumMath.nextRandomNumber() * Math.PI - Math.PI/2.0;
+    var x = radius * Math.cos(theta) * Math.cos(phi);
+    var y = radius * Math.sin(phi);
+    var z = radius * Math.sin(theta) * Math.cos(phi);
+    return new Cartesian3(x, y, z);
+}
+
+function randomFunction(position) {
+    void(position); // Unused parameter
```
Comment (pjcozzi) [inline generator/lib/createPointCloudTile.js]: Is this to pass JSHint?  There might be a JSHint comment to disable this warning in this function like we do for other things.

File: generator/lib/util.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+'use strict';
```
Comment (pjcozzi) [inline generator/lib/util.js]: This isn't a public API, but still name this fil utility.js.

**Comment by pjcozzi** [comment]
At quick glance, the code looks good, really clean.  Is there anything you want me to look at carefully?

**Comment by pjcozzi** [comment]
Also, no need to overbuild here, let's get this reasonable and merge, and roadmap anything that isn't essential.

**Comment by pjcozzi** [comment]
@lilleyse is this ready?

**Comment by lilleyse** [comment]
No, I'll finish this up today.

**Comment by pjcozzi** [comment]
> No, I'll finish this up ~~today~~ soon.

lol.  Bump.

**Comment by lilleyse** [comment]
This is ready pending an npm release to gltf-pipeline.

**Comment by lilleyse** [comment]
These tilesets are still gltf 1.0 based. It's probably still worth updating Cesium in the meantime as these fix several `3d-tiles` branch test failures.

**Comment by pjcozzi** [comment]
>  It's probably still worth updating Cesium in the meantime as these fix several 3d-tiles branch test failures.

Sounds good, just use separate data PRs since we'll want to squash commit history for them or put the test tilesets in cesium.com (maybe even download them as part of npm install or similar config?).

**Comment by pjcozzi** [comment]
Did I do something wrong:
```
pcozzi@C1200906-00 /c/source/3d-tiles-tools/generator ((d464b59...))
$ npm run test

> 3d-tiles-generator@0.1.0 test c:\source\3d-tiles-tools\generator
> gulp test

[06:51:22] Using gulpfile c:\source\3d-tiles-tools\generator\gulpfile.js
[06:51:22] Starting 'test'...
Started
Spec started
F
  3d-tiles-generator
    × runs
      - Failed: 3d-tiles-generator.js exited with an error code of 1
      - Error: Timeout - Async callback was not invoked within timeout specified
 by jasmine.DEFAULT_TIMEOUT_INTERVAL.



Failures:
1) 3d-tiles-generator runs
  Message:
    Failed: 3d-tiles-generator.js exited with an error code of 1
  Stack:
    Error: Failed: 3d-tiles-generator.js exited with an error code of 1
        at ChildProcess.<anonymous> (c:\source\3d-tiles-tools\generator\specs\bi
n\generatorSpec.js:17:17)
        at ChildProcess.g (events.js:260:16)
        at emitTwo (events.js:87:13)
        at ChildProcess.emit (events.js:172:7)
  Message:
    Error: Timeout - Async callback was not invoked within timeout specified by
jasmine.DEFAULT_TIMEOUT_INTERVAL.
  Stack:
    Error: Timeout - Async callback was not invoked within timeout specified by
jasmine.DEFAULT_TIMEOUT_INTERVAL.
        at Timer.listOnTimeout (timers.js:92:15)

1 spec, 1 failure
Finished in 5.062 seconds

**************************************************
*                    Failures                    *
**************************************************

1) 3d-tiles-generator runs
  - Failed: 3d-tiles-generator.js exited with an error code of 1
  - Error: Timeout - Async callback was not invoked within timeout specified by
jasmine.DEFAULT_TIMEOUT_INTERVAL.

Executed 1 of 1 spec (1 FAILED) in 5 secs.
[06:51:27] Finished 'test' after 5.12 s
```

Note that even with a clean clone, I had to `git checkout origin/generator`, not `git checkout generator`.

**Comment by lilleyse** [comment]
Do you get an error if you just run `node bin/3d-tiles-generator.js`?

It's possible the test is timing out since it runs the entire generator.

**Comment by pjcozzi** [comment]
Both of these pass on my Mac.  Do you think this has anything to do with Windows?  Otherwise, can we change the timeout?

**Comment by lilleyse** [comment]
Both pass on Windows for me. I will change the timeout to be longer.

**Comment by lilleyse** [comment]
Updated. Default timeout of 5 seconds was too slow, bumped to 20 seconds (usually finishes in 4-6 for me)

**Pull request merged**